<h1 style='color:blue;' align='center'>KFold Cross Validation</h2>

In [1]:
import numpy as np

import matplotlib.pyplot as plt

from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier

from sklearn.datasets import load_digits

digits = load_digits()

In [7]:
X = digits.data
X

array([[ 0.,  0.,  5., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ..., 10.,  0.,  0.],
       [ 0.,  0.,  0., ..., 16.,  9.,  0.],
       ...,
       [ 0.,  0.,  1., ...,  6.,  0.,  0.],
       [ 0.,  0.,  2., ..., 12.,  0.,  0.],
       [ 0.,  0., 10., ..., 12.,  1.,  0.]])

In [8]:
y = digits.target
y

array([0, 1, 2, ..., 8, 9, 8])

In [9]:
X.shape

(1797, 64)

In [15]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

### Logistic Regression

In [11]:
lr = LogisticRegression(solver='liblinear', multi_class='ovr')
lr.fit(X_train, y_train)
lr.score(X_test, y_test)

0.9518518518518518

In [16]:
lr = LogisticRegression(solver='liblinear', multi_class='ovr')
lr.fit(X_train, y_train)
lr.score(X_test, y_test)

0.9555555555555556

### SVM

In [5]:
svm = SVC(gamma='auto')
svm.fit(X_train, y_train)
svm.score(X_test, y_test)

0.26851851851851855

In [14]:
svm = SVC(gamma='auto')
svm.fit(X_train, y_train)
svm.score(X_test, y_test)

0.35

### Random Forest

In [6]:
rf = RandomForestClassifier(n_estimators=40)
rf.fit(X_train, y_train)
rf.score(X_test, y_test)

0.9685185185185186

In [17]:
rf = RandomForestClassifier(n_estimators=40)
rf.fit(X_train, y_train)
rf.score(X_test, y_test)

0.9722222222222222

### train_test_split ---> Iteration : 01
 - Logistic Regression : First Cell ---> Accuracy : 0.9518518518518518
 - SVM : First Cell ---> Accuracy : 0.26851851851851855
 - Random Forest : First Cell ---> Accuracy : 0.9685185185185186
 

### train_test_split ---> Iteration : 02
 - Logistic Regression : First Cell ---> Accuracy : 0.9555555555555556
 - SVM : First Cell ---> Accuracy : 0.35
 - Random Forest : First Cell ---> Accuracy : 0.9722222222222222
 
 
#### For each iteration of **train_test_split** accuracy changed.

-----------------------------

<h2 style='color:purple'> 1. KFold cross validation</h2>

### a. Basic example

In [18]:
from sklearn.model_selection import KFold
kf = KFold(n_splits=3)
kf

KFold(n_splits=3, random_state=None, shuffle=False)

In [19]:
for train_index, test_index in kf.split([1,2,3,4,5,6,7,8,9]):
    print(train_index, test_index)

[3 4 5 6 7 8] [0 1 2]
[0 1 2 6 7 8] [3 4 5]
[0 1 2 3 4 5] [6 7 8]


### b. Use KFold for our digits example

In [20]:
def get_score(model, X_train, X_test, y_train, y_test):
    model.fit(X_train, y_train)
    return model.score(X_test, y_test)

In [21]:
from sklearn.model_selection import StratifiedKFold
folds = StratifiedKFold(n_splits=3)

scores_logistic = []
scores_svm = []
scores_rf = []

for train_index, test_index in folds.split(digits.data,digits.target):
    X_train, X_test, y_train, y_test = digits.data[train_index], digits.data[test_index], \
                                       digits.target[train_index], digits.target[test_index]
    scores_logistic.append(get_score(LogisticRegression(solver='liblinear',multi_class='ovr'), X_train, X_test, y_train, y_test))  
    scores_svm.append(get_score(SVC(gamma='auto'), X_train, X_test, y_train, y_test))
    scores_rf.append(get_score(RandomForestClassifier(n_estimators=40), X_train, X_test, y_train, y_test))

In [22]:
scores_logistic

[0.8948247078464107, 0.9532554257095158, 0.9098497495826378]

In [23]:
scores_svm

[0.3806343906510851, 0.41068447412353926, 0.5125208681135225]

In [24]:
scores_rf

[0.9215358931552587, 0.9382303839732888, 0.9248747913188647]

<h2 style='color:purple'> 2. cross_val_score function</h2>

In [25]:
from sklearn.model_selection import cross_val_score

**Logistic regression model performance using cross_val_score**

In [26]:
cross_val_score(LogisticRegression(solver='liblinear', multi_class='ovr'), digits.data, digits.target, cv=3)

array([0.89482471, 0.95325543, 0.90984975])

**svm model performance using cross_val_score**

In [27]:
cross_val_score(SVC(gamma='auto'), digits.data, digits.target, cv=3)

array([0.38063439, 0.41068447, 0.51252087])

**random forest performance using cross_val_score**

In [28]:
cross_val_score(RandomForestClassifier(n_estimators=40), digits.data, digits.target, cv=3)

array([0.92654424, 0.94824708, 0.92487479])

cross_val_score uses stratifield kfold by default

<h2 style='color:purple'> 3. Parameter tunning using k fold cross validation</h2>

In [29]:
scores1 = cross_val_score(RandomForestClassifier(n_estimators=5), digits.data, digits.target, cv=10)
np.average(scores1)

0.8764773432650529

In [30]:
scores2 = cross_val_score(RandomForestClassifier(n_estimators=20), digits.data, digits.target, cv=10)
np.average(scores2)

0.944891371818746

In [36]:
scores3 = cross_val_score(RandomForestClassifier(n_estimators=30), digits.data, digits.target, cv=10)
np.average(scores3)

0.9494121957589801

In [31]:
scores4 = cross_val_score(RandomForestClassifier(n_estimators=40), digits.data, digits.target, cv=10)
np.average(scores4)

0.9476753569211669

Here we used cross_val_score to
fine tune our random forest classifier and figured that having around 40 trees in random forest gives best result. 

----------------------

### Reference:

https://www.youtube.com/watch?v=gJo0uNL-5Qw&ab_channel=codebasics